### ISSUE TITLE
Searching in skill-tree not as comprehensive as in-game

### ISSUE THREAD

**Issue by Ovalidem** [PathOfBuildingCommunity/PathOfBuilding#2173]
When I search in Path of Building, I get less results than when I search in the skill-tree in-game. I think this is because PoB searches for the whole string, instead of in a different way.

Meaning, when I search "evasion armour" (without the quotation marks) in-game, I get all nodes containing any of those two words, like the Keystone Iron Reflexes:
![in-game search](https://user-images.githubusercontent.com/10993412/107643984-1f7b1480-6c77-11eb-9305-19b45c828634.png)

But if I had searched the same thing in PoB, I get nothing (the red bubbles are mine, not PoB's) because there's no string in any of the nodes containing "evasion armour" (no quotes):
![PoB search](https://user-images.githubusercontent.com/10993412/107644191-69fc9100-6c77-11eb-9f68-e38a6a444149.png)

When I make builds, I try to look for specific combinations of things. Say that in this example, I wanted evasion and armour. I can't search "Evasion Armour" in PoB, because no nodes show up, so I end up having to search through either all the "evasion" searched nodes which also have armor or vice versa.

Would it be possible to update the way searching functions in PoB? It would immensely improve the flow with which I can make builds.

**Comment by Ethrel** [comment]
I see where and what would need to change (I think). I think I see a workable solution to this. However, before bringing time and effort to the table to work up a pr, I'd like to hear thoughts from the repo maintainers (and possibly the rest of the community) regarding this.

I can definitely see how the behavior differs, however I'm not sure that the change in behavior in the PoB search is largely desired.

Essentially, it would change from an exact string match to a split on whitespace and fuzzy match on each word, creating an or search (does node match X or Y). Is it worth implementing this change? Should there be some form of toggle between behaviors? I see no reason the search behaviors would have to be mutually exclusive, but toggle-able behavior would increase complexity of use for the end user (not that the PoB curve is especially low to begin with).

Anyone have any thoughts on specific implementation details?

**Comment by Ovalidem** [comment]
The only argument I can give for implementing this, and it's easy to ask for something when I'm not the one having to implement it, is that when you're looking for specifics things you need to add in your build, you can search for those nodes directly.
I am not very experienced in the late-game or making whole entire builds / theory-crafting and then implementing them, but the way I make builds is literally picking stuff I might encounter while leveling. So if I might need say, armour and elemental resistances, I look for nodes that have them. As it is right now, I either have to only put in either "Armour" or "Elemental" and manually pick between the nodes that pop up.

TL;DR: it would make searching for specific nodes faster and less labour intensive.

**Comment by Wires77** [comment]
@Ethrel This is a worthy change, and you're right in that this new behavior doesn't need a toggle or be mutually exclusive with the old search behavior.  One thing that splitting on spaces will change is if someone is looking for a multi-word mod, the "OR" search behavior will find it and potentially many more mods, so if it's not too complicated, we can have quotes bypass the split on spaces.

Technically, in-game the tree search does an "AND" search in order to find nodes that match both words, not just one of them, so I think implementing it that way would be more accurate.

**Comment by Ethrel** [comment]
@Wires77 You're right. It is indeed an AND. I'd like to think I'd have caught that as I was working it up...if I do it, I do plan on checking results between tree and in-game anyway.

Quotes to bypass the split on spaces shouldn't be too hard. Just a little extra string manipulation. Which, tbh, should be relatively minor, providing I make sure to do that manipulation AFTER verifying the search cache. On top of that, PoE search actually does this too. I just had to load up the game to verify, but PoE search does indeed maintain string integrity on quoted search terms.
